src/extension/ncwms/src/main/java/au/org/emii/geoserver/wms/Ncwms.java

```diff
@@ -0,0 +1,304 @@
+            System.out.println(queryString);
+
+            URL wmsUrl = new URL(wmsUrlStr + "?" + queryString);
+
+            IOUtils.copy(wmsUrl.openConnection().getInputStream(), response.getOutputStream());
+        }
+        catch (Exception e) {
+            LOGGER.log(Level.SEVERE, String.format("Problem proxying url '%s'", url));
+            e.printStackTrace();
+        }
+    }
+
+    private String encodeMapForRequest(Map<String, String[]> params) {
+        StringBuilder sb = new StringBuilder();
+        try {
+            for (Map.Entry<String, String[]> param : params.entrySet()) {
+                if (sb.length() != 0) sb.append('&');
+                sb.append(URLEncoder.encode(param.getKey(), StandardCharsets.UTF_8.name()));
+                sb.append('=');
+                sb.append(URLEncoder.encode(param.getValue()[0], StandardCharsets.UTF_8.name()));
+            }
+        }
+        catch (Exception e) {
+            LOGGER.log(Level.SEVERE, String.format("Error encoding parameters: '%s'", e.getMessage()));
+        }
+
+        return sb.toString();
+    }
+
+    private List<String> getSupportedStyles() {
+        return new ArrayList<String>() {{
+            add("barb");
+            add("fancyvec");
+            add("trivec");
+            add("stumpvec");
+            add("linevec");
+            add("vector");
+            add("boxfill");
+        }};
+    }
+
+    private List<String> getPalettes() {
+        return new ArrayList<String>() {{
+            add("redblue");
+            add("alg");
+            add("greyscale");
+            add("alg2");
+            add("ncview");
+            add("occam");
+            add("rainbow");
+            add("sst_36");
+            add("ferret");
+            add("occam_pastel-30");
+        }};
+    }
+
+    private static String cqlFilterForTimestamp(String timestamp, String timeFieldName) {
+        LOGGER.log(Level.INFO, String.format("Returning cql for timestamp '%s'", timestamp));
+        return String.format("%s = %s", timeFieldName, timestamp);
+    }
+
+    private static String cqlFilterForSameDay(String day, String timeFieldName) {
+        String timeStart = day;
+        String timeEnd = getNextDay(timeStart);
+        LOGGER.log(Level.INFO, String.format("Returning times of day '%s'", day));
+
+        String cqlFilter = String.format(
+            "%s >= %s AND %s < %s",
+            timeFieldName, timeStart,
+            timeFieldName, timeEnd
+        );
+
+        return cqlFilter;
+    }
+
+    private static String getWfsUrlParameters(String wfsLayer, String propertyName, String cqlFilter) throws UnsupportedEncodingException {
+        String urlParameters =
+            String.format(
+                "typeName=%s&SERVICE=WFS&outputFormat=csv&REQUEST=GetFeature&VERSION=1.0.0&PROPERTYNAME=%s",
+                wfsLayer, propertyName
+            );
+
+        if (cqlFilter != null && cqlFilter != "") {
```

Comment:
I think an empty string is an empty string, in any encoding. Probably I could use `cqlFilter.empty()`, which might be better.